packages/edit-navigation/src/store/actions.js

```diff
@@ -49,169 +38,307 @@ export function setSelectedMenuId( menuId ) {
+	registry,
+	dispatch,
+} ) => {
+	const lock = await registry
+		.dispatch( 'core' )
+		.__unstableAcquireStoreLock( STORE_NAME, [ 'savingMenu' ], {
+			exclusive: true,
+		} );
 	try {
-		// Save edits to the menu, like the menu name.
-		yield dispatch(
-			'core',
-			'saveEditedEntityRecord',
-			'root',
-			'menu',
-			menuId
-		);
+		const menuId = post.meta.menuId;
 
-		const error = yield select(
-			'core',
-			'getLastEntitySaveError',
-			'root',
-			'menu',
-			menuId
-		);
+		await registry
+			.dispatch( 'core' )
+			.saveEditedEntityRecord( 'root', 'menu', menuId );
+
+		const error = registry
+			.select( 'core' )
+			.getLastEntitySaveError( 'root', 'menu', menuId );
 
 		if ( error ) {
 			throw new Error( error.message );
 		}
 
-		// Save blocks as menu items.
-		const batchSaveResponse = yield* batchSave(
-			menuId,
-			menuItemsByClientId,
-			post.blocks[ 0 ]
+		// Batch save menu items
+		const oldMenuItems = await dispatch(
+			resolveSelectMenuItems( post.meta.menuId )
+		);
+		const newMenuItems = await dispatch(
+			computeNewMenuItems( post, oldMenuItems )
+		);
+		await dispatch(
+			batchSaveDiff( 'root', 'menuItem', oldMenuItems, newMenuItems )
 		);
-
-		if ( ! batchSaveResponse.success ) {
-			throw new Error( batchSaveResponse.data.message );
-		}
 
 		// Clear "stub" navigation post edits to avoid a false "dirty" state.
-		yield dispatch(
-			'core',
-			'receiveEntityRecords',
-			NAVIGATION_POST_KIND,
-			NAVIGATION_POST_POST_TYPE,
-			[ post ],
-			undefined
-		);
+		await registry
+			.dispatch( 'core' )
+			.receiveEntityRecords(
+				NAVIGATION_POST_KIND,
+				NAVIGATION_POST_POST_TYPE,
+				[ post ],
+				undefined
+			);
 
-		yield dispatch(
-			noticesStore,
-			'createSuccessNotice',
-			__( 'Navigation saved.' ),
-			{
+		await registry
+			.dispatch( noticesStore )
+			.createSuccessNotice( __( 'Navigation saved.' ), {
 				type: 'snackbar',
-			}
-		);
+			} );
 	} catch ( saveError ) {
 		const errorMessage = saveError
 			? sprintf(
 					/* translators: %s: The text of an error message (potentially untranslated). */
 					__( "Unable to save: '%s'" ),
 					saveError.message
 			  )
-			: __( 'Unable to save: An error ocurred.' );
-		yield dispatch( noticesStore, 'createErrorNotice', errorMessage, {
-			type: 'snackbar',
-		} );
+			: __( 'Unable to save: An error o1curred.' );
+		await registry
+			.dispatch( noticesStore )
+			.createErrorNotice( errorMessage, {
+				type: 'snackbar',
+			} );
+	} finally {
+		await registry.dispatch( 'core' ).__unstableReleaseStoreLock( lock );
 	}
-} );
+};
+
+const computeNewMenuItems = ( post, oldMenuItems ) => async ( {
+	dispatch,
+} ) => {
+	const entityIdToBlockId = await dispatch(
+		getEntityRecordIdToBlockIdMapping( post.id )
+	);
+	const blockIdToOldEntityRecord = Object.fromEntries(
+		oldMenuItems
+			.map( ( entityRecord ) => [
+				entityIdToBlockId[ entityRecord.id ],
+				entityRecord,
+			] )
+			.filter( ( [ blockId ] ) => blockId )
```

Comment:
Yeah I looked at after a day and had a hard time too. I replaced it with a good, old for loop. I didn't add any comments for now because I couldn't come up with anything better than "map block ids to their corresponding entity records" and the variable name is `blockIdToOldEntityRecord`.

Reply:
Huh I actually don't think this is hard to read to be honest. I'd imagine `reduce` to be worse, but maybe that's just me. Though I agree for-loop would be better in general.

Reply:
A loop seems like a good compromise, thanks 👍